__tests__/setup-go.test.ts

```diff
@@ -0,0 +1,299 @@
+import * as tc from '@actions/tool-cache';
+import * as core from '@actions/core';
+import fs = require('fs');
+import osm = require('os');
+import path = require('path');
+import {run} from '../src/main';
+import * as httpm from '@actions/http-client';
+import * as im from '../src/installer';
+import * as sys from '../src/system';
+import {ITypedResponse} from '@actions/http-client/interfaces';
+
+let goJsonData = require('./data/golang-dl.json');
+
+describe('setup-go', () => {
+  let inputs = {} as any;
+  let os = {} as any;
+
+  let inSpy: jest.SpyInstance;
+  let findSpy: jest.SpyInstance;
+  let cnSpy: jest.SpyInstance;
+  let logSpy: jest.SpyInstance;
+  let getSpy: jest.SpyInstance;
+  let platSpy: jest.SpyInstance;
+  let archSpy: jest.SpyInstance;
+  let dlSpy: jest.SpyInstance;
+  let exSpy: jest.SpyInstance;
+  let cacheSpy: jest.SpyInstance;
+
+  beforeEach(() => {
+    // @actions/core
+    inputs = {};
+    inSpy = jest.spyOn(core, 'getInput');
+    inSpy.mockImplementation(name => inputs[name]);
+
+    // node 'os'
+    os = {};
+    platSpy = jest.spyOn(osm, 'platform');
+    platSpy.mockImplementation(() => os['platform']);
+    archSpy = jest.spyOn(osm, 'arch');
+    archSpy.mockImplementation(() => os['arch']);
+
+    // @actions/tool-cache
+    findSpy = jest.spyOn(tc, 'find');
+    dlSpy = jest.spyOn(tc, 'downloadTool');
+    exSpy = jest.spyOn(tc, 'extractTar');
+    cacheSpy = jest.spyOn(tc, 'cacheDir');
+    getSpy = jest.spyOn(im, 'getVersions');
+
+    // writes
+    cnSpy = jest.spyOn(process.stdout, 'write');
+    logSpy = jest.spyOn(console, 'log');
+    getSpy.mockImplementation(() => <im.IGoVersion[]>goJsonData);
+    cnSpy.mockImplementation(line => {
+      // uncomment to debug
+      // process.stderr.write('write:' + line + '\n');
+    });
+    logSpy.mockImplementation(line => {
+      // uncomment to debug
+      // process.stderr.write('log:' + line + '\n');
+    });
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+    jest.clearAllMocks();
```

Comment:
@bryanmacfarlane From [the docs](https://jestjs.io/docs/en/mock-function-api.html#mockfnmockreset), it looks like reset does everything clear does and more

Do you need to call `jest.restoreAllMocks` during `afterAll`, which "restores the original (non-mocked) implementation."